### ISSUE TITLE
"space-before-blocks" not working correctly

### ISSUE THREAD

**Issue by Hudda** [eslint/eslint#13553]
**Tell us about your environment**

* **ESLint Version: v7.6.0 (Currently used)**
* **Node Version: v12.16.2**
* **npm Version: v6.14.4**

**What parser (default, Babel-ESLint, etc.) are you using?**
@typescript-eslint/parser

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
"space-before-blocks": [
      "error",
      "always"
    ],
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
cacheSkillRights(skillId: string, skillRights: SkillRights): void{
    this.skillRightsCache[skillId] = cloneDeep(skillRights);
  }
```

```bash

```

**What did you expect to happen?**
Eslint raise 'space-before-blocks' error as there are no space after void and before '{'

**What actually happened? Please include the actual, raw output from ESLint.**
Eslint passes successfully

**Are you willing to submit a pull request to fix this bug?**
No

**Comment by anikethsaha** [comment]
Thanks for the report. 

The rule is working fine. There is a missing space between `void`  and `{`, as the rule clearly stats a space before the blocks.
Also, It looks like there should be a rule overriding this rule in `@typescript-eslint/eslint-plugin` for better TS experience.
Currently, there is no rule for this use case in `typescript-eslint` that provides TS support.

So maybe it can be a rule request to [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint).

**Comment by Hudda** [comment]
> The rule is working fine. There is a missing space between void and {, as the rule clearly stats a space before the blocks.

I think there is a misunderstanding, eslint didn't report any error, but it should because the space is missing.

**Comment by anikethsaha** [comment]
Eslint didnt report because it doesnt know about types. So it won't work properly with typescript. I think `typescript-eslint` should support this rule. 

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgc3BhY2UtYmVmb3JlLWJsb2NrczogW1wiZXJyb3JcIiwgXCJhbHdheXNcIl0qL1xuXG5mdW5jdGlvbiBjYWNoZVNraWxsUmlnaHRzKHNraWxsSWQsIHNraWxsUmlnaHRzKXtcbiAgICB0aGlzLnNraWxsUmlnaHRzQ2FjaGVbc2tpbGxJZF0gPSBjbG9uZURlZXAoc2tpbGxSaWdodHMpO1xufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsiaW5kZW50IjoyfSwiZW52Ijp7fX19)

**Comment by mdjermanovic** [comment]
[space-before-blocks](https://eslint.org/docs/rules/space-before-blocks) doesn't control spaces between a keyword and `{`:

> This rule ignores spacing which is between a keyword and a block. The spacing is handled by the keyword-spacing rule

Maybe the [@typescript-eslint/keyword-spacing](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/keyword-spacing.md) rule could be updated to support this case, if it doesn't already.

**Comment by mdjermanovic** [comment]
It's interesting that the rule works well with `string`, `number`, and other types because they have `Identifier` tokens.

`void` just happens to be a keyword as well. In this context, it's more a type identifier than a keyword, but its token is still `Keyword`.

We added  [this check](https://github.com/eslint/eslint/blob/6677180495e16a02d150d0552e7e5d5f6b77fcc5/lib/rules/space-before-blocks.js#L91) to avoid conflicts with other rules, but there should be no conflicts in this particular case because `keyword-spacing` checks spacing after `void` in unary expressions only.

Wondering if we could change the rule to just not check the type of the previous token if the given block is the body of a function declaration/function expression.

**Comment by anikethsaha** [comment]
in this line https://github.com/eslint/eslint/blob/6677180495e16a02d150d0552e7e5d5f6b77fcc5/lib/rules/space-before-blocks.js#L102

if we change the following, 
```diff
- if (precedingToken && !isConflicted(precedingToken) && astUtils.isTokenOnSameLine(precedingToken, node)) {
+ if (precedingToken && (['FunctionDeclaration', 'FunctionExpression'].includes(node.type) || !isConflicted(precedingToken)) && astUtils.isTokenOnSameLine(precedingToken, node)) {
```

the rule will be working fine I suppose, (tests are passing) but still, there would be parsing error for types (using the default parser).

**Comment by mdjermanovic** [comment]
Yes, we could add a fixture parser test. I'd also like to fix some things about reporting, and refactor the code to avoid the `ASTNode|Token` logic in `checkPrecedingSpace` since it's a bit confusing and error-prone.

**Comment by anikethsaha** [comment]
Yeah, 
One thought, why it is a bug? it is intended behavior, right? the parser is working fine just that the code is in TS so it is supposed to break. Are there any cases that is not being tested because of the conflict check and that is not covered in arrow and keyword spacing rules?

or maybe I am missing something?

**Comment by mdjermanovic** [comment]
This isn't technically a bug, the rule works fine with JS code.

Core rules may or may not work well with TS code. Most of the rules work perfectly well, those that have some complex issues with TS code are usually extended in the typescript plugin.

The idea here is to just slightly change the logic where it doesn't affect JS code and doesn't look unreasonable for JS code, but will make the rule work well with TS code. We used to make such small changes before.

**Comment by anikethsaha** [comment]
Coool,

**Comment by Hudda** [comment]
Hi guys, so what is the conclusion here, will you implement the functionality? Thanks!

**Comment by Hudda** [comment]
@anikethsaha any update on this?

**Comment by anikethsaha** [comment]
I am in support of the change mentioned by @mdjermanovic (that would fix this issue as well. ) .

**Comment by mdjermanovic** [comment]
Sorry for the delay! I believe the intention of the keyword check was to avoid conflicts with the `keyword-spacing` rule, but there is no need to do that for function body blocks. I'll work on this.

**Comment by mdjermanovic** [comment]
Marking this issue and the PR accepted as there are three :+1: from team members.
